File: Introduction to Data Structures: Array and String/2. Introduction to 2D Array/Spiral Matrix/SpiralMatrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:

        # edge case if matrix is empty return an empty array
        if len(matrix) == 0:
            return []

        # initialize an empty array to store the results
        result = []

        # find the top most row
        minRow = 0

        # find the bottom most row
        maxRow = len(matrix) - 1

        # find the left most column
        minCol = 0

        # find the right most column
        maxCol = len(matrix[0]) - 1

        # while the number of elements needed > 0
        while minRow <= maxRow and minCol <= maxCol:

            # go right until we reach the edge
            for col in range(minCol, maxCol+1):

                # append element to the list
                result.append(matrix[minRow][col])

            # remove the top edge
            minRow += 1

            # go down until we reach the bottom edge
            for row in range(minRow, maxRow+1):

                # append element to the list
                result.append(matrix[row][maxCol])

            # remove the right edge
            maxCol -= 1

            # do a check to see if minRow <= maxRow before proceeding
            if (minRow <= maxRow):

                # go left until we reach the left edge
                for col in range(maxCol, minCol-1, -1):

                    # append element to the list
                    result.append(matrix[maxRow][col])

                # remove the bottom edge
                maxRow -= 1

            # do another check to see if minCol <= maxCol
            if (minCol <= maxCol):

                # go up until you reach min row
                for row in range(maxRow, minRow-1, -1):

                    # appened element to list
                    result.append(matrix[row][minCol])

                # remove left edge
                minCol += 1

        # return the result array
        return result

File: Introduction to Data Structures: Array and String/2. Introduction to 2D Array/Spiral Matrix/SpiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        
        // initialize an empty ArrayList to store the result
        List <Integer> result = new ArrayList(); 
        
        // edge case: occurs when the length of input is 0
        if (matrix.length == 0) {
            return result;
        }
        
        // create variables to represent left, right, top, and bottom boundaries
        int leftCol = 0;
        int rightCol = matrix[0].length - 1; 
        int topRow = 0;
        int bottomRow = matrix.length - 1;
        
        // while the left boundary is less than or equal to the right boundary
        // while the top boundary is less than or equal to the bottom boundary
        while (leftCol <= rightCol && topRow <= bottomRow) {
        
            // move to the right
            for (int col = leftCol; col <= rightCol; col++) {
            
                // add element to result
                result.add(matrix[topRow][col]);
            }
            
            // remove top boundary from calculation
            topRow++;
            
            // move down
            for (int row = topRow; row <= bottomRow; row++) {
        
                // add element to result
                result.add(matrix[row][rightCol]);
            }
            // remove right boundary from calculation
            rightCol--; 
            
            // check if the top row <= the bottom row 
            if (topRow <= bottomRow) {
        
                // if it is go left
                for (int col = rightCol; col >= leftCol; col--) {
        
                    // add element to result
                    result.add(matrix[bottomRow][col]);
                }
            }
            
            // remove bottom boundary from calculation
            bottomRow--;
            
            // check if the left boundary <= right boundary
            if (leftCol <= rightCol) {
        
                // if it is go up
                for (int row = bottomRow; row >= topRow; row--) {
        
                    // add element to result
                    result.add(matrix[row][leftCol]);  
                }    
            }
            // remove left boundary from calculation
            leftCol++;
        }
        // return result
        return result;
    }
}